### DISCUSSION TITLE
How to set ContactsContract.RawContacts.SOURCE_ID on insert?

### DISCUSSION THREAD

**Discussion by marrale** [vestrel00/contacts-android#299] [Q&A] [answered]
In my previous code I'm using the "ContactsContract.RawContacts.SOURCE_ID" Field to store my contact primary key provided from backend. How can I achieve this with this library?
My need is to have a contact primary key (string) to be able to later find the contact and edit/delete them. I didn't found the way to write a contact with my custom "lockupKey".
Thanks Alessandro

**Comment by vestrel00** [comment]
Hey! This is a great callout! Currently, the library does not support setting or reading the `ContactsContract.SyncColumns.SOURCE_ID` for RawContacts. This seems like your use case is more for manipulating the contacts provider from a sync adapter perspective more than an application perspective (this library has minimal to no support for sync adapter specific use cases). However, we can definitely add it in 😄 

Just to make sure we are in the same page about the purpose of this field, here's what I know about it; https://developer.android.com/reference/android/provider/ContactsContract.RawContacts

<img width="857" alt="Screenshot 2023-08-28 at 6 48 06 AM" src="https://github.com/vestrel00/contacts-android/assets/4332753/45bd35f0-14df-4d44-baa6-1b0956d8d933">

There are two use cases where the source ID is set,

1. App creates new RawContact (source ID is null at this point) set to an Account with sync adapter X  -> sync adapter X detects (or is notified) that  a new RawContact with a null source ID and notifies the corresponding remote sync server/service/backend of the new RawContact -> backend creates a new entry server side and returns the source ID -> sync adapter X sets the source ID of the existing RawContact to the value returned by the backend.
2. There is a new RawContact in the backend that the device does not have (because it was created elsewhere) -> sync adapter X creates a new RawContact received from the backend, which includes the RawContact source ID.

In other words, we need to be able to set the `ContactsContract.RawContacts.SOURCE_ID` on insert and update. I tried both and it works even without setting `ContactsContract.CALLER_IS_SYNCADAPTER`  to true. This implies that the source ID can be set by apps too and not just sync adapters, though I would highly advise apps against setting it or changing its value.

>  I didn't found the way to write a contact with my custom "lockupKey".

I believe you mean the Contact "lookup key", which is automatically set by the Contacts Provider. The value of the Contact's lookup key typically uses the RawContact's source id (if it is not null) as a part of a larger string. I did find that setting an existing RawContact's source id results in an **automatic** change to the Contact's lookup key.

**Reply by marrale** [reply]
Thanks for your answer and searching. Yes, the purpose of my app is to do crud operations on app contacts retrieved from a backend (sync adapter), this why I need the Source ID field. 
I found out that setting the "SOURCE_ID" ensures that the "Lookupkey" always consists of 5 characters + "SOURCE_ID". 
the 5 characters seem to be the same at least per "insert" operation (4 numbers and one letter).
Accordingly, one can search for "SOURCE_ID" within the "Lookupkey". Specifically: Contact.LookupKey.contains("MY_SOURCE_ID") or Contact.LookupKey.endsWith("MY_SOURCE_ID"). Both should work fine for searching and deleting contacts via your API if you have set the "SOURCE_ID" manually.
But have the possiblity to directly read the source Id would be a cleaner way.

**Reply by vestrel00** [reply]
#### Note to self. 

Here is an example of a newly created RawContact associated with a `com.google` Account **prior** to syncing,

```
 D  #### Contacts table
 D  Contact id: 31, lookupKey: 2059r31-38462A40563E3A4436
 D  #### RawContacts table
 D  RawContact id: 31, contactId: 31, sourceId:null
```

After sync,

```
 D  #### Contacts table
 D  Contact id: 31, lookupKey: 2059i7b975d4a8fec684e
 D  #### RawContacts table
 D  RawContact id: 31, contactId: 31, sourceId:7b975d4a8fec684e
```

For a RawContact that is not associated with an Account (null),

```
 D  #### Contacts table
 D  Contact id: 32, lookupKey: 0r32-3032543A2E324644405A
 D  #### RawContacts table
 D  RawContact id: 32, contactId: 32, sourceId:null
```

This is not important for this discussion. Just thought I might as well leave this note for future reference in case I need it 😄

**Reply by marrale** [reply]
Thanks for this interesting example ;-)

**Comment by vestrel00** [comment, marked as answer]
@marrale, I created an issue for adding support for your use case; #300. It will be included in the next release ([0.3.1](https://github.com/vestrel00/contacts-android/discussions/288)).

In the meantime, you should still be able to use this library right now to make your life easier. If you want to adopt this library now instead of waiting for the next release, you can do this in order to insert a new RawContact with the source id provided from your backend...

1. [Insert a new RawContact into your Account](https://vestrel00.github.io/contacts-android/basics/insert-contacts/) with all the data (name, emails, phones, etc) from your backend **using this library**. The RawContact's source id will be null at this point. 
2. Then, update the source id of the newly inserted RawContact **using your own code**. 

Or, to make sure a new RawContact is not created in the backend side when a new RawContact is created locally without a source ID, you can instead...

1. Insert new RawContact into your Account u**sing your own code** with the non-null source ID from your backend, without any data rows.
2. Then, [update the newly inserted RawContact](https://vestrel00.github.io/contacts-android/basics/update-contacts/) with all the data (name, emails, phones, etc) from your backend **using this library**.

Regardless of whether you are or will be using this library, thanks for starting this discussion and look forward to #300 🙏 ❤️

**Reply by marrale** [reply]
Thanks for including the SOURCE_ID  as described in https://github.com/vestrel00/contacts-android/issues/300 when creating/update and search for the SOURCE_ID at RAW_CONTACTS (where SOURCE_ID equals "my source id") in the next Version.

**Reply by vestrel00** [reply]
@marrale, I completed #300. Once the next library version ([0.3.1](https://github.com/vestrel00/contacts-android/discussions/288)) is released, you will be able to do the following...

Insert new RawContact with sourceId,

```kotlin
Contacts(this)
        .insert()
        .rawContact { 
            sourceId = "mycustomsourceid"
        }
        .commit()
```

Update sourceId of existing RawContact,

```kotlin
Contacts(this)
        .update()
        .rawContacts(rawContact.mutableCopy {
            sourceId = "mycustomsourceid"
        })
        .commit()
```

Query using the sourceId field for matching,

```kotlin
Contacts(this)
        .rawContactsQuery()
        .rawContactsWhere(accounts, RawContactsFields.SourceId equalTo "mycustomsourceid")
        .find()
```

Delete using sourceId field for matching,

```kotlin
Contacts(this)
        .delete()
        .rawContactsWhere { SourceId equalTo "mycustomsourceid" }
        .commit()
```

**Reply by marrale** [reply]
@vestrel00 thanks for adding the requests, this prefect for my needs.

**Reply by vestrel00** [reply]
🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 
This is now release in [0.3.1](https://github.com/vestrel00/contacts-android/releases/tag/0.3.1)!
🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥

**Reply by marrale** [reply]
Thanks @vestrel00 for adding this new useful feature!

**Comment by vestrel00** [comment]
By the way, I also went ahead and created an issue for supporting CRUD operations on Group's SOURCE_ID; #303. It's pretty much the same as what has been discussed here, except instead of RawContacts, it is Groups.
